### PR TITLE
Use raw time string from DB to generate ActiveRecord#cache_version

### PR DESCRIPTION
**Pull request by schneems**
Currently, the `updated_at` field is used to generate a `cache_version`. Some database adapters return this timestamp value as a string that must then be converted to a Time value. This process requires a lot of memory and even more CPU time. In the case where this value is only being used for a cache version, we can skip the Time conversion by using the string value directly.

## Details

- This PR preserves existing cache format by converting a UTC string from the database to `:usec` format.
- Some databases return an already converted Time object, in those instances, we can directly use `created_at`.
- The `updated_at_before_type_cast` can be a value that comes from either the database or the user. We only want to optimize the case where it is from the database.
- If the format of the cache version has been changed, we cannot apply this optimization, and it is skipped.
- If the format of the time in the database is not UTC, then we cannot use this optimization, and it is skipped.

Some databases (notably PostgreSQL) returns a variable length nanosecond value in the time string. If the value ends in a zero, then it is truncated For instance instead of `2018-10-12 05:00:00.000000` the value `2018-10-12 05:00:00` is returned. We detect this case and pad the remaining zeros to ensure consistent cache version generation.

### Memory

Before: Total allocated: 743842 bytes (6626 objects)
After: Total allocated: 702955 bytes (6063 objects)

(743842 - 702955) / 743842.0 # => 5.4% ⚡️⚡️⚡️⚡️⚡️

### Speed

Using the CodeTriage application and derailed benchmarks this PR shows between 9-11% (statistically significant) performance improvement versus the commit before it.

## Thanks

Special thanks to @lsylvester for helping to figure out a way to preserve the usec format and for helping with many implementation details.

### CODE REVIEW

**Comment by schneems** [comment]
I think we're in the clear since all the data from usec has to come from the raw string in the first place. I'm happy to dig in more with other verification methods if you can think of any. cc/ @matthewd

**Comment by sgrif** [comment]
Does this change result in `model.updated_at = model.updated_at` changing the cache key? Do we care if so?

**Comment by schneems** [comment]
> Does this change result in model.updated_at = model.updated_at changing the cache key? Do we care if so?

Interesting. I didn't think about assignment changing the cache key. I think if you have `model.updated_at = Time.now` then the cache key will change based on current behavior but not with this patch. For `model.updated_at = model.updated_at` I don't think the cache key will change since it will still use the string before typecast so modifying `updated_at` shouldn't affect anything.

**Comment by matthewd** [comment]
ISTM the greater type-mismatch concern would be that the cache key must match between a freshly-written object (with only a "real" Time) and that same object when later loaded from the database.

5% is pretty huge, though: I see no reason that shouldn't be easily demonstrable on a real request-sized benchmark, without needing to lean on the [still undemonstrated to me] "memory not allocated = time saved" theory.

**Comment by schneems** [comment]
>  without needing to lean on the [still undemonstrated to me] "memory not allocated = time saved" theory.

From my earlier PR we see a [20% reduction in objects and a 20% reduction in page load time](https://github.com/rails/rails/pull/33758#issuecomment-417689955). 

I ran the numbers for this PR and I did find some relatively surprising results running the page speed for this:

Before: 13.560740   0.962090  14.555312 ( 30.606835)
After: 11.225604   0.900667  12.159477 ( 25.296171)
Diff: (30.606835 - 25.296171) / 30.606835 # => **17.35 % faster**

Which makes sense since the parsing and generating of time objects is CPU intensive (in addition to being memory intensive). Also that time parsing was already memory optimized by some other PRs I had.

Note that i'm not advocating for `memory not allocated = time saved` i'm advocating for `memory not allocated ~ time saved` as it's a better proxy for repeatability. I trust my memory numbers, they're  consistent. 

For raw speed numbers to be truly accurate I would need to reproduce the above test dozens/hundreds of times calculate variance and use something like a "students T test" to actually have faith in the number. Each of those dozens of runs (per case) takes 30 seconds and then I have to repeat them for the hypothesis and the null hypothesis case. Multiply that by the changes that get requested (usually) for a PR and that pattern is not be very sustainable.

**Comment by brodock** [comment]
That gives me one idea: do you think it would be possible to typecast model attributes when it is used for the first time in opposite of typecasting in advance as I believe it is today?

**Comment by schneems** [comment]
@brodock that is the current behavior. It’s done on demand which is why this PR works by not using that field if it’s not needed.

File: activerecord/lib/active_record/integration.rb
Comment (rafaelfranca) [inline activerecord/lib/active_record/integration.rb]: Will this change the format of the cache key and thus invalidate all the cache entries?

**Comment by rafaelfranca** [comment]
I have to ask, but are you bechmarking on macos?

File: activerecord/lib/active_record/integration.rb
Comment (sgrif) [reply activerecord/lib/active_record/integration.rb]: I don't think we've exposed a public API for checking individual attributes, but `@attributes["updated_at"].has_been_read?` would do it.

**Comment by schneems** [comment]
> I have to ask, but are you bechmarking on macos? And with Puma in a forked process?

Yes i'm on a mac. I'm using derailed_benchmarks without a server so i'm using the raw rack interface directly essentially i'm manually `.call`-ing the app. https://github.com/schneems/derailed_benchmarks/blob/cd6befa4aa7d7d10e6fbbd126ff60fb4ec8cf2f2/lib/derailed_benchmarks/tasks.rb#L104-L108

You can try the benchmark yourself here is my branch of codetriage for Rails 6 https://github.com/codetriage/codetriage/tree/schneems/6.ohhh that i've been perf testing against. 

First clone then make sure you're pointing to a relative copy of rails in the gemfile. Then

```
export RAILS_LOG_TO_STDOUT=1
export RAILS_SERVE_STATIC_FILES=1
export SECRET_KEY_BASE=foobarlol
export DEVISE_SECRET_KEY=1
RAILS_ENV=production bundle exec rake db:create:all db:migrate db:seed
```

Now you can run perf tests

```
bundle exec derailed exec perf:test
```

I git checkout this branch and run the perf tests, then checkout the commit right before  my commit and re-run.

Also clear your screen between runs as all logs go to stdout and your terminal buffer might have an impact. Also close EVERY app on your machine. Ideally turn off the wifi and walk away with nothing running other than this script.

The number I gave earlier (17%) isn't statistically valid it's from two runs and is likely skewing to both sides of a natrual distribution for both the before and after number. If you're interested in an actual number I can stand behind then i'll need some more time.

**Comment by dhh** [comment]
We definitely cannot change the cache key format in a point release, unless
there's a bullet proof fallback, such that we don't invalidate all caches
unexpectedly for someone in production.

On Mon, Sep 10, 2018 at 2:27 PM, Richard Schneeman <notifications@github.com

**Comment by rafaelfranca** [comment]
Ok, if you are in MacOS there is a chance that https://bugs.ruby-lang.org/issues/14435 is on the play here, but given it is not forking I find it unlikely. I'd still try to see if 2.6.0 trunk make a difference here.

**Comment by sgrif** [comment]
Alternative to trunk would be to use something like apache bench, which is probably the most appropriate tool for a full stack benchmark. Even the 5% memory savings is a bit surprising to me. Is this hitting an action using an array of hundreds of records for its cache key? There just shouldn't be enough casts happening here to really add up.

@dhh Since this is targeting master, the cache key invalidation would be happening in 6.0. Is it acceptable to do that in a major version in your opinion? The other big concern is that this will cause any patch which changes the database representation of the timestamp (such as switching to binary representation on PG which I've been wanting to do) to also invalidate all existing cache keys.

**Comment by dhh** [comment]
I think this is an acceptable change for 6.0, as long as we validate that
the preliminary results are actually legit on likely production
hardware/OS. Because it's definitely a hassle, and you'll have to plan for
it, but if there's a 5% performance bonus, that seems reasonable.

I think it's fine that changing the column type of the key used for the
cache key invalidates cache keys. Just need to document that fact.

**Comment by schneems** [comment]
On CodeTriage it's hitting the main page which is I think 50 cached fragments in each backed by an AR object. The allocations are happening mostly in `time_value.rb` you can see it come up several times in my "top" output. The `/type/helpers/time_value.rb:72` is my 6th largest source of memory allocations by bytes.


```
# bytes allocated | location of allocation
allocated memory by location
-----------------------------------
    122491  .../rails/activesupport/lib/active_support/core_ext/string/output_safety.rb:172
     49288  .../codetriage/app/views/layouts/application.html.slim:1
     49248  .../codetriage/app/views/layouts/_app.html.slim:1
     24904  .../codetriage/app/views/pages/_repos_with_pagination.html.slim:1
     18984  .../ems/rack-mini-profiler-1.0.0/lib/patches/db/pg.rb:92
     17343  .../rails/activemodel/lib/active_model/type/helpers/time_value.rb:72 # <===================
     17056  .../rails/actionview/lib/action_view/digestor.rb:23
     14920  .../ems/rack-mini-profiler-1.0.0/lib/patches/db/pg.rb:9
     14720  .../rails/activesupport/lib/active_support/core_ext/object/to_query.rb:43
     14400  .../rails/activerecord/lib/active_record/result.rb:150
     12000  .../rails/activemodel/lib/active_model/type/helpers/time_value.rb:80 # <===================
     10944  .../rails/activesupport/lib/active_support/cache.rb:599
      9600  .../rails/activemodel/lib/active_model/attribute_set/builder.rb:25
      8815  .../ems/rack-mini-profiler-1.0.0/lib/mini_profiler/storage/file_store.rb:27
      8000  .../rails/activemodel/lib/active_model/attribute.rb:9
      8000  .../rails/activerecord/lib/active_record/attribute_methods.rb:281
      7904  .../rails/actionpack/lib/abstract_controller/caching/fragments.rb:92
      6656  .../rails/actionview/lib/action_view/helpers/cache_helper.rb:228
      5600  .../rails/activerecord/lib/active_record/persistence.rb:81
      4558  .../rails/activesupport/lib/active_support/cache/memory_store.rb:123
      4400  .../rails/activemodel/lib/active_model/attribute_set/builder.rb:16
      4320  .../rails/actionpack/lib/action_dispatch/routing/route_set.rb:199
      4300  .../rails/activemodel/lib/active_model/type/helpers/time_value.rb:64 # <===================
```

**Comment by dhh** [comment]
At Basecamp, we definitely have pages where you can end up hitting 100+
cached fragments backed by an AR object on a cold boot. But then the
Russian doll kicks in, so that the majority of the time you'd hit far less
than that. Still, wouldn't mind improving that cold scenario, or the case
where people aren't able to use Russian dolls for some reason. As long as
we don't fuck with ergonomics or precision.

On Mon, Sep 10, 2018 at 2:45 PM, Richard Schneeman <notifications@github.com

**Comment by sgrif** [comment]
Interesting. It's definitely an allocation hungry cast, but it's shocking to me that it could end up being that large of a portion of allocation. I would be incredibly shocked if this had a measurable impact on actual response time though, and would like to see some confirmation of results if we're making the change on those grounds. Even if it doesn't affect response time though, 5% in overall allocated memory per request is worth it on its own, as long as we acknowledge that this only affects applications which:

1. Are using a large number of Active Record objects as cache keys
2. Are not otherwise accessing `updated_at` (which they wouldn't for cache hits -- I suspect this really only affects warm cache cases on index pages, where you check dozens of records individually even on a warm cache)

With all of that said, if we're fine with invalidating existing caches in 6.0, I think I'm ok with this change.

**Comment by lsylvester** [comment]
@schneems I assume that you are running this issuing PostgreSQL, which retrieves the values from the dbs as strings. For other adapters like Mysql2, the values are received as Time objects. There still would be savings in not having to convert to `ActiveSupport::TimeWithZone`, but it would need explicit formatting to include units more precise than seconds.

**Comment by schneems** [comment]
Yes this is with postgres. I believe the PG gem is now capable of providing type cast values now, perhaps an alternative could be to use the pg type cast for this one attribute for now. I'm not sure about the performance of that time casting versus the one directly in Rails but presumably it's faster and uses less memory. If we go that route then we could perhaps get all the savings without needing to bust the cache key. How hard would that be @sgrif ?

**Comment by schneems** [comment]
Okay I have some statistically significant results for actual performance numbers. Here's the raw output https://gist.github.com/schneems/0277b3d42b6414dbbe69188f2ca46408

On average the delta between my patch and master (the commit right before my change) is **11.50%**. The standard deviations for both sets are well under the average delta and when I applied a students-t test for statistical validity I got a number so small it's not showing as anything other than 0. I am EXTREMELY confident in saying that this improvement comes based on my patch. Note that the exact number may vary based on what else the computer is doing and the alignment of the sun and moon. I can say with confidence that i've seen a 11.5% performance gain with this patch, but I cannot claim with certainty that is the exact improvement it provides (i.e. YMMV but it's 100% for sure going to be faster).

You can reproduce the results on my `schneems/6.ohhh` branch that I linked to earlier by running this command.

```
$ BEFORE_BRANCH=schneems/faster_cache_version AFTER_BRANCH=dd29fabebf1b702031213361b7757be5892cd59f bundle exec rake compare_branches
```

It assumes that the directory codetriage is in is the same directory that your rails folder is in. It also assumes that you have my patch it the `schneems/faster_cache_version` branch.

**Comment by sgrif** [comment]
@schneems The PG gem's behavior doesn't match Rails unfortunately, but date parsing is going to be slow no matter who's doing it. The best source of a performance win would be to use binary encoding, which is trivial for us to parse (but that's a much larger change that isn't feasible right now)

File: activerecord/lib/active_record/integration.rb
Comment (schneems) [reply activerecord/lib/active_record/integration.rb]: I guess having different behavior based on the status of the object would lead to cache bloat as sometimes the key would be different values. So that's not a great idea. I'm mostly trying to brainstorm on ways to minimize the cache invalidation issue. Having talked this one through with you i don't think it's a good path forward.

Comment (sgrif) [reply activerecord/lib/active_record/integration.rb]: I think to go forward with this change we need to accept that it will invalidate existing caches.

**Comment by schneems** [comment]
I tested this against mysql by changing my adapter to `mysql2` and re-running my memory benchmarks. This is what I see:

Before: Total allocated: 741754 bytes (6977 objects)
After: Total allocated: 716055 bytes (6624 objects)
Diff:  (741754 - 716055) / 741754.0 # 3.46 % fewer bytes allocated

While it's not as dramatic as the postgres change it is less memory allocation and a 3.46% diff is still pretty good. You can see my app I used in the branch:  https://github.com/codetriage/codetriage/tree/schneems/6.ohhh-mysql

Sounds like we're mostly back to square one. This patch will make things faster for sure but will invalidate also caches.

Do we pull this trigger or is there something else we want to investigate first?

**Comment by lsylvester** [comment]
It should be possible to manually format the raw strings from the database into the format required for the cache_version without much overhead and object allocation, but it might need to be database specific and we would need to know that *_before_type_cast was set from the database. For `pg`, it would just be `updated_at_before_type_cast.delete('- :.')`, which would only be allocating a single string.

**Comment by sgrif** [comment]
@lsylvester And then when we switch to binary format for dates?

**Comment by matthewd** [comment]
Invalidating caches in a minor release is as acceptable as in a major: in both cases, it's ideally avoided, but okay if justified. (And it seems like we end up invalidating at least a good chunk for each new release series lately.)

So while it should be a conscious decision, I don't think that's a showstopper. My remaining concern is around the fact that `_before_type_cast` contains the value.. before type casting, and that that's not _necessarily_ going to match the form in which it would come out of the database. @sgrif caches aside, would something like `value_for_database` manage to give us the existing/original DB string if available, but also have the best chance of accurately recreating it if needed?

File: activerecord/lib/active_record/integration.rb
Comment (ohaddahan) [inline activerecord/lib/active_record/integration.rb]: @schneems might be able to squeeze a bit more out of it by replacing `try` with `respond_to?`.
https://gist.github.com/ohaddahan/f54f9f891aa2bf6c4dc83c00e4936f86

`respond_to?` on an attribute that exists 463K vs. `try` 362K , that's 27% increase.
On a non existing attribute `respond_to?` and `try` are within error maring at 350K.

Comment (ohaddahan) [reply activerecord/lib/active_record/integration.rb]:
```
if respond_to?(:updated_at_before_type_cast)
    updated_at_before_type_cast
  elsif respond_to?(:updated_at)
    updated_at.utc.to_s(:usec)
  end
```

**Comment by sgrif** [comment]
@matthewd `value_for_database` will round trip through type casting. We could in theory specialize it in `FromDatabase` to just return `value_before_type_cast`, but I think we intentionally avoided that to help us ensure that `type.serialize(type.deserialize(x))` is always an identity function.

Honestly, I think it's fine if we end up with different cache keys depending on where the object came from. It's uncommon to use an object as a cache key where its values came from anywhere *other* than the database. Having a view where `updated_at` *sometimes* comes from the db, and *sometimes* comes from the user, and expecting them to share a cache key seems unlikely enough to me that I think we can ignore it.

I do agree with you that we need to normalize the behavior in the case of `updated_at` coming from the user though. We should probably only use `updated_at_before_type_cast` if `!updated_at_came_from_user?`

**Comment by sgrif** [comment]
> Invalidating caches in a minor release is as acceptable as in a major

Other folks in this thread seem to disagree with that, we should probably make it clear what our position is.

File: activerecord/lib/active_record/integration.rb
Comment (sgrif) [reply activerecord/lib/active_record/integration.rb]: Please avoid commenting on open PRs with "have you tried X?" unless you have benchmark numbers to back it up

Comment (ohaddahan) [reply activerecord/lib/active_record/integration.rb]: You mean like the gist link I posted? @sgrif

Comment (sgrif) [reply activerecord/lib/active_record/integration.rb]: @ohaddahan We generally don't accept microbenchmarks, and this PR is discussing the context of a full request/response cycle.

Comment (schneems) [reply activerecord/lib/active_record/integration.rb]: We talked about this offline. Thanks for the comment! Though I want to keep this PR to one change at a time. I think we would need to show that the above change affected the total time of execution for a request in a measurable way.

Comment (ohaddahan) [reply activerecord/lib/active_record/integration.rb]: @sgrif in the case where `update_at_before_type_cast` exist, we get a 13% improvement.
https://gist.github.com/ohaddahan/2579c1472be753a17a6a310bc5f83c8a
On the `else` case the call to `utc.to_s(:usec)` is too dominate to get significant benefit from the change to `respond_to?`.
Keep in mind that I didn't run the testing suite, so I'm assuming we don't need to propagate down stream, which is the only reason to use `try` over `respond_to?` in this case.
Also, I'm using the latest `try` from your latest PR, but if you think switching one method isn't worth 13% increase, it's your call.

Comment (ohaddahan) [reply activerecord/lib/active_record/integration.rb]: @schneems didn't meant to insert it into this PR, just a heads up to be considered as an additional PR.

**Comment by schneems** [comment]
> Other folks in this thread seem to disagree with that, we should probably make it clear what our position is.

Since this change is for a major version bump do we need to do that in this thread? I think this is as good a place as any, but I also kinda want this merged in 😅.

I think that...it depends. If we introduce binary format and have to coerce that back into this format, that will likely be a -10% perf hit right? If we're willing to break the cache to get a 10% faster response here I think I would be willing to break my app cache again to not have to pay another -10% cache later. That being said I'm not running at basecamp or shopify scale.

Is there anything that can be done to mitigate the rollout of something like this for a company "at scale"? I.e. gradual rollout on servers so by the time it's at 100% then hopefully the majority of the cache isn't cold? Are there considerations like "thundering herd" we have to worry about where all the machines try to update all the same cache keys at the same time? Anything else?

Maybe we could document mitigation strategies and do a better job advertising when a change does break the cache, for example when we post on https://blog.rubyonrails.org, use the dreaded post-install message, or some other way.

I do think we want to be careful about cache keys changing. Perhaps as a place to start, we can say:

- Major: X.0.0 - Yes, cache keys can change in major versions, but need a reason.
- Minor: 6.X.0 - Maybe, cache key changes are eligible on minor versions but need more burden of proof of benefit than major version bump.
- Patch: 6.0.X - No, cache keys cannot change in a patch update.

What do you think?

**Comment by sgrif** [comment]
@schneems 

> Since this change is for a major version bump do we need to do that in this thread? I think this is as good a place as any, but I also kinda want this merged in 😅.

Yes, I think we do. As I've been saying, this increases the surface area of changes we could make which will invalidate existing caches. Since the consensus seemed to be that we are only willing to do that in major version bumps, we should address the problem of "everyone who can review a PR has to be aware that any change which addresses the database representation of datetimes cannot be made in minor versions". This isn't an obvious thing to reviewers even if we're familiar with the history, and I think we need to either somehow enforce that this doesn't accidentally happen, or be clear that we're fine with invalidating caches in minor versions. If we're not willing to invalidate caches in minor versions, this change also restricts our ability to do things like change to binary representation for a performance bump in a minor version.

> I think that...it depends. If we introduce binary format and have to coerce that back into this format, that will likely be a -10% perf hit right?

It'll be whatever the cost of casting is (2 `Date` allocations with the APIs available today) plus the cost of the formatting method to whatever specific format we're expecting.

> I think I would be willing to break my app cache again to not have to pay another -10% cache later. That being said I'm not running at basecamp or shopify scale.

Right, that's the problem -- We're making that call for all our users. I personally think we should not be expecting cache keys to remain consistent across Rails versions, but it's not my call to make.

> Maybe we could document mitigation strategies and do a better job advertising when a change does break the cache

I'd like to see us at least have an official policy on this, and some sort of automated test enforcing that.

**Comment by dhh** [comment]
I don't have any problem breaking the cache format in a major version. Of
course, it has to be widely known that it's happening, and the pay-off has
to be material. For a minor version bump, I'd feature-flag it. Retain the
existing format for existing apps, unless you actively change a setting,
but all new apps get the best performance.

But our next version from master is 6.0, so in this particular case we
don't really have to worry about any mitigation strategy.

We just have to be clear and sure that when we make the code more
convoluted for performance reasons that there's a substantial payoff. I
like the work we've done in that department on this PR!

**Comment by sgrif** [comment]
In that case my main concern is just the increased surface area for invalidated caches, and making sure we are aware if we break them in the future. "This subtly changes the db representation of a datetime for performance reasons" feels too likely to slip in unexpectedly

**Comment by lsylvester** [comment]
I think that this change will reduce the precision of the cache versions when the database connection is returning time objects. Using Mysql2, before this change I was seeing versions of `20180912233441000000` (it is mysql5.6 which is why there are no milliseconds, but newer versions would have values other than 0 at the end), but with this change I am seeing value of `2018-09-13 09:38:30 +1000` or `2018-09-12 23:34:41 UTC` depending if the updated at is from the db or user, which doesn't have any precision beyond seconds. @schneems Can we add some tests to ensure the `usec` component of the time is included in the version so we don't accidentally drop this.,

I think that we need to check the type of the value that is returned from `updated_at_before_type_cast`, and format it to a string if it is not already a string. 

If it is a string we could `.delete('- :.')` which would give us the same value as if we has formatted it (at least for PostgreSQL, but that would no longer be required once we are using binary encoding. Also, if the database is in `local` time instead of `utc` then they won't match - so in that case we might need to parse and reformat the versions, but that would only effect a small percentage of apps.

**Comment by schneems** [comment]
> Using Mysql2, before this change I was seeing versions of 20180912233441000000 

Looking at the code for `usec` it is in `core_ext/time/conversions.rb`: 

```
    usec: "%Y%m%d%H%M%S%6N",
```

So yes, those zeroes at the end are essentially just padding and it is limited to 1 second accuracy, even with type cast since the source string is limited to that accruacy. Do you have an example when the string does not have microsecond data but the `to_s(:usec)` does? If that's the case where does the extra precision come from if it's not in the string?

> I think that we need to check the type of the value that is returned from updated_at_before_type_cast, and format it to a string if it is not already a string.

It gets cast to a string by the cache store already. If we're given a DateTime object from the user then we could convert it in some other custom way than calling `to_s` on it, but we would have to match all the formats for all the databases to be truly compatible both ways. How common is it that a user will manually set updated_at? How likely is it that they'll do it before touching the cache? If existing adapters already have logic for date formatting then maybe we could use that?

**Comment by schneems** [comment]
@sgrif regarding policies and better ways of testing maybe @rafael or @matthewd have more thoughts.

**Comment by schneems** [comment]
@sgrif and friends. This PR and connected issue might be of interest https://github.com/rails/rails/pull/34005

**Comment by lsylvester** [comment]
@schneems MySQL does support sub second times, but they need have the columns created with `precision` option as by default it is 0. There are tests for this at https://github.com/rails/rails/blob/bd107964197989b5b3ec3395e3ab011bb9d662d7/activerecord/test/cases/integration_test.rb#L156 but they are only for checking the cache key supports sub-second precision, and not that the cache version does. As they a generated via different code paths - the sub-second support for `cache_version`s is currently untested.

I will try to get a PR up soon to add this testing in so we can verify if these changes have any impact of sub-second cache version support for MySQL.

**Comment by lsylvester** [comment]
@schneems I have made the PR #34089 which passes for me on master but fails against this branch.

While doing this I realised that the `cache_version` is `to_param`ed before being used - so one option would be to update `Time#to_param` (or replace `to_param` with a more specialised method) to match the desired format in the cache key and then `updated_at_before_type_cast` would work regardless if the value from the database is a `String` or a `Time`.

**Comment by schneems** [comment]
Thanks for the test case @lsylvester ! I took a look and can confirm it's failing on this branch

```
Failure:
IntegrationTest#test_cache_version_format_is_not_too_precise [/Users/rschneeman/Documents/projects/rails/activerecord/test/cases/integration_test.rb:191]:
--- expected
+++ actual
@@ -1,2 +1 @@
-# encoding: US-ASCII
-"2018-10-06 17:06:59 UTC"
+"2018-10-06 17:06:59.090853" # <== Raw
```

However, it doesn't look like the cause is lack of precision from the database. The issue (at least with sqlite) is that in some cases the value is a string and in other cases, it is a Time object. This is the issue that @sgrif was talking about where `updated_at_before_type_cast` can come from the DB or from the user (after someone did an assignment).

I think the solution here would be to manually format the Time object to match the raw format from the DB.

**Comment by lsylvester** [comment]
I am also seeing this failure using mysql2

```
Failure:
IntegrationTest#test_cache_version_format_is_precise_enough [/Users/lachlansylvester/Repos/rails/rails/activerecord/test/cases/integration_test.rb:181]:
Expected # encoding: US-ASCII
"2018-10-06 21:21:53 UTC" to not be equal to # encoding: US-ASCII
"2018-10-06 21:21:53 UTC".


bin/test test/cases/integration_test.rb:173
```

**Comment by schneems** [comment]
I'm not able to really dig in tonight. But it looks like the same issue at heart. Both entries from the non-typecast method call are Time objects in that test so when you call to_s on them without formatting it gives you an entry without subsecond timing. 

Your test is merged into master. It's good behavior with or without this patch.

**Comment by lsylvester** [comment]
Once solution for this is to define `Time#to_param` to that when a time is returned from the `cache_version` it gets correctly formatted as per this commit https://github.com/lsylvester/rails/commit/816bc770086f31f3e548711bde0e17127a74a76a. But I think that the `to_param` method (and format) might be a bit generic of a name and could possibly conflict other uses of it and cause regressions if anyone is currently using `Time#to_param` in their apps. 

Let me know if you are happy with this approach and I can add some tests and get a PR happening.

**Comment by schneems** [comment]
I pulled in the tests from @lsylvester on master and got them passing. To do that I changed the behavior so that we only use the `updated_at_before_type_cast` when it is a raw string. When it is a time object we default to `updated_at`.  Tests are currently passing as of https://github.com/rails/rails/pull/33835/commits/0c2fee70445c4e9959adb06bd047aa3a96135324. 

I also standardized the format to match with the raw string coming back from the database. That change required a number of tests to be updated that were asserting the format of the cache key.

I think this is ready for another round of reviews. There are tests that exercise this functionality but there are currently no tests that guarantee that this method was called and used. I'm open to adding something like that, for now I wanted the minimal acceptable set of changes that will get master passing.

File: activerecord/lib/active_record/integration.rb
Comment (lsylvester) [inline activerecord/lib/active_record/integration.rb]: I am not sure that checking if the timestamp is a String is sufficient - if a user set it to a String that wan't in the desired format then the cache_key wouldn't match before and after save. 

In addition database adapters that perform the type casting would no longer get any benefit from this as it will be a `Time` that will get cast to `ActiveSupport::TimeWithZone` and back as part of the the formatting of the time.

I think that it is worth checking if the value is from the database as part of this optimisation.

**Comment by schneems** [comment]
Any idea how to check if it came from the DB or not?

**Comment by lsylvester** [comment]
Also, to revisit the breaking cache version format - it is possible to keep the format be directly transforming the string from the database.

I have a microbenchmark for this at https://gist.github.com/lsylvester/32da6af3dc0a117bfc282e8834f89256 - and it shows that keeping the format would be about 1.2x slower than this implementation instead of the 4x slower of implementation before this change.

```
Calculating -------------------------------------
                 New    292.214k (± 9.0%) i/s -      1.460M in   5.033789s
  New keeping format    250.280k (± 8.5%) i/s -      1.255M in   5.045426s
            Original     69.712k (± 5.5%) i/s -    349.223k in   5.024369s

Comparison:
                 New:   292214.1 i/s
  New keeping format:   250279.6 i/s - same-ish: difference falls within error
            Original:    69711.7 i/s - 4.19x  slower
```

so we could get most of the benefits of this without breaking the format. Just something to keep in mind.

**Comment by schneems** [comment]
@lsylvester I updated my test to use a hash. That's a great idea to preserve existing cache key format. I love it! I updated this branch. I also added the guard check for ensuring that the DB is set to `:utc` and I moved all the checks into their own private method as the line was getting really long.

While implementing these changes I noticed that my memory crept back up with the addition of the `!try(:updated_at_came_from_user?)` check. That lead me to find another memory improvement in `respond_to?` for Active Record (Here's the PR https://github.com/rails/rails/pull/34197). I would appreciate if you all could take a look there. I would like to get that merged in so I can do another benchmark of performance with the extra guard checks and the string mutation.

**Comment by lsylvester** [comment]
@schneems I think that we could bypass `try` altogether - we are calling it up to tree times here - and each one depends on the updated_at being present.

If we returned early unless `has_attribute?('updated_at')` the calls to try (and respond_to?) would no longer be required.

Otherwise - those changes look good - will make more comments in that PR.

**Comment by schneems** [comment]
I updated to use the `has_attribute?` check. Which looks better and removes a few `try` calls (3 in the worst case). So this patch is no longer dependent on my other perf PR (though #34197 is still worth it).

With this checking logic i'm able to see a (very) statistically significant perf gain of 9.79%. The absolute number here may vary a bit, but there is no doubt that this patch is still fast.

I also discovered an issue with an intermittent failing test. It turns out that postgres (and mybe other adapters) will truncate trailing zeros in the nanosecond field. For instance instead of `"2018-10-12 05:00:00.000000"` the value `"2018-10-12 05:00:00"` is returned. When that happens we can detect by checking the length and padding with zeros.

**Comment by schneems** [comment]
PR is rebased, I left the silent failure commit separate so it can be cherry-picked if we decide not to merge this.

File: activerecord/lib/active_record/integration.rb
Comment (lsylvester) [inline activerecord/lib/active_record/integration.rb]: Can we use `ljust` here - I think it reads better and is slightly faster according to my benchmarks (not that this is a common path to be in).

```suggestion
          key.ljust(20, "0")
```

File: activerecord/test/cases/cache_key_test.rb
Comment (lsylvester) [inline activerecord/test/cases/cache_key_test.rb]: Do these records need to be saved to the db. I think we should be able to check these behaviours on new records without saving and reloading them.

Comment (schneems) [reply activerecord/test/cases/cache_key_test.rb]: In the `CacheMeWithVersion.create` the object assigned to `updated_at` is a Time object. That triggers the `timestamp.is_a?(String)` check which means we skip the "fast" cache behavior so then `updated_at` is guaranteed to be called. 

There might be another way to trigger this behavior that I've not found. This testing logic seemed the closest to the actual behavior I was trying to test (values generated in user space skip my logic, values from the DB use the new logic).

If you want to play around with it, the fastest way to run this test file is running this inside the AR directory:

```
bundle exec rake test:sqlite3_mem  TEST=test/cases/cache_key_test.rb
```

Comment (lsylvester) [reply activerecord/test/cases/cache_key_test.rb]: @schneems For the tests that are ready values from the database I think that saving the records is required - but for the tests where values are being set by the user (as a String, Time or Hash) the tests could would with unsaved records. See https://github.com/lsylvester/rails/commit/c4ff214c741bda3e5d9957edc4bba2bb7d7c49a8

Comment (schneems) [reply activerecord/test/cases/cache_key_test.rb]: The reason I didn't want to go with something like:

```ruby
record = CacheMeWithVersion.new
assert_called(record, :updated_at) do
  record.updated_at = Time.now.to_s
  record.cache_version
end
```

Is if the line we're testing gets removed:

```
record = CacheMeWithVersion.new
assert_called(record, :updated_at) do
  # record.updated_at = Time.now.to_s
  record.cache_version
end
```

Then the test would still pass, since the value is coming from the user already, as opposed to being from the DB. Essentially without the value being from the db it's possible to make a mistake in modifying test or in modifying the code and have all the tests still pass.

**Review by lsylvester — APPROVED** [review]

**Pull request merged**
